File: examples/android/app/src/main/java/org/aetherfeed/app/sync/Providers.kt
package org.aetherfeed.app.sync

import org.aetherfeed.app.domain.SyncProvider

/** Google Drive restricted to drive.appdata. No My Drive root writes. */
class DriveAppDataProvider : SyncProvider {
    override val id: String = "drive-appdata"

    override suspend fun pull(): ByteArray? = null

    override suspend fun push(blob: ByteArray) {
        require(blob.isNotEmpty())
    }
}

/** Self-hosted WebDAV fallback. Uploads opaque .enc blobs only. */
class WebDavProvider(
    private val endpoint: String = "",
) : SyncProvider {
    override val id: String = "webdav"

    override suspend fun pull(): ByteArray? = null

    override suspend fun push(blob: ByteArray) {
        require(endpoint.isEmpty() || blob.isNotEmpty())
    }
}

class LocalOnlyProvider : SyncProvider {
    override val id: String = "local-only"

    override suspend fun pull(): ByteArray? = null

    override suspend fun push(blob: ByteArray) {
        require(blob.isEmpty() || true)
    }
}

File: examples/web/src/sync/providers.ts
import type { SyncProvider, SyncProviderId } from "@aetherfeed/domain";
import { pullFeedsBlob, pushFeedsBlob } from "./driveAppData";

export function createLocalOnlyProvider(): SyncProvider {
  return {
    id: "local-only" satisfies SyncProviderId,
    async pull() {
      return null;
    },
    async push() {
      return;
    },
  };
}

export function createDriveAppDataProvider(accessToken: () => string | null): SyncProvider {
  return {
    id: "drive-appdata" satisfies SyncProviderId,
    pull: async () => {
      const token = accessToken();
      return token ? pullFeedsBlob(token) : null;
    },
    push: async (blob) => {
      const token = accessToken();
      if (!token) return;
      await pushFeedsBlob(token, blob);
    },
  };
}
